subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileSystemSnapshotter.java

```diff
@@ -208,20 +224,190 @@ public void appendToHasher(BuildCacheHasher hasher) {
     }
 
     private class FileVisitorImpl implements FileVisitor {
-        private final List<FileSnapshot> fileTreeElements;
+        private final static int BATCH_SIZE = 32;
+        private final ArrayList<FileSnapshot> fileTreeElements;
+        private final Runnable[] buffer;
+        private int bufferSize;
+        private boolean completed;
 
-        FileVisitorImpl(List<FileSnapshot> fileTreeElements) {
-            this.fileTreeElements = fileTreeElements;
+        FileVisitorImpl() {
+            this.fileTreeElements = Lists.newArrayList();
+            this.buffer = new Runnable[BATCH_SIZE];
         }
 
         @Override
         public void visitDir(FileVisitDetails dirDetails) {
-            fileTreeElements.add(new DirectoryFileSnapshot(getPath(dirDetails.getFile()), dirDetails.getRelativePath(), false));
+            fileTreeElements.add(new DirectoryFileSnapshot(internPath(dirDetails.getFile()), dirDetails.getRelativePath(), false));
         }
 
         @Override
-        public void visitFile(FileVisitDetails fileDetails) {
-            fileTreeElements.add(new RegularFileSnapshot(getPath(fileDetails.getFile()), fileDetails.getRelativePath(), false, fileSnapshot(fileDetails)));
+        public void visitFile(final FileVisitDetails fileDetails) {
+            if (buffer != null) {
+                visitFileBuffered(fileDetails);
+            } else {
+                visitFileDirect(fileDetails);
+            }
+        }
+
+        private void visitFileBuffered(final FileVisitDetails fileDetails) {
+            final DeferredFileSnapshot deferred = new DeferredFileSnapshot(fileDetails);
+            buffer[bufferSize++] = deferred;
+            fileTreeElements.add(deferred);
+            if (bufferSize == BATCH_SIZE) {
+                flush();
+            }
+        }
+
+        private void visitFileDirect(final FileVisitDetails fileDetails) {
+            fileTreeElements.add(new RegularFileSnapshot(internPath(fileDetails.getFile()), fileDetails.getRelativePath(), false, fileSnapshot(fileDetails)));
+        }
+
+        public List<FileSnapshot> getElements() {
+            if (completed) {
+                return fileTreeElements;
+            }
+            flush();
+            int i = 0;
+            for (FileSnapshot element : fileTreeElements) {
+                if (element instanceof DeferredFileSnapshot) {
```

Comment:
Yes, I hesitated with `ArrayList<Object>`, but I think with the final implementation it would make sense. I didn't use a separate list on purpose, to avoid allocating yet another buffer.